packages/backend/src/open_payments/auth/middleware.ts

```diff
@@ -33,39 +34,47 @@ export function createAuthMiddleware({
         return
       }
       const authService = await ctx.container.use('authService')
-      const grant = await authService.introspect(token)
-      if (!grant || !grant.active) {
+      const tokenInfo = await authService.introspect(token)
+      if (!tokenInfo || !tokenInfo.active) {
         ctx.throw(401, 'Invalid Token')
       }
       if (
-        !grant.includesAccess({
+        !tokenInfo.includesAccess({
           type,
           action,
           identifier: ctx.paymentPointer.url
         })
       ) {
         ctx.throw(403, 'Insufficient Grant')
       }
+      try {
+        if (!(await verifySigAndChallenge(tokenInfo.key.jwk, ctx))) {
+          ctx.throw(401, 'Invalid signature')
+        }
+      } catch (e) {
+        ctx.status = 401
+        ctx.throw(401, `Invalid signature: ${e.message}`)
+      }
       await GrantReference.transaction(async (trx: Transaction) => {
```

Comment:
Kind of as an aside, does it make sense to directly reference the model here, in the middleware? Should this live inside the `grantService` instead?

Reply:
- https://github.com/interledger/rafiki/issues/703